src/compas/datastructures/assembly/part.py

```diff
@@ -0,0 +1,187 @@
+        The local coordinate system of the part.
+    shape : :class:`compas.geometry.Shape`
+        The base shape of the part geometry.
+    features : list of tuple(:class:`compas.geometry.Shape`, str)
+        The features added to the base shape of the part geometry.
+    transformations : deque of :class:`compas.geometry.Transformation`
+        The stack of transformations applied to the part geometry.
+        The most recent transformation is on the left of the stack.
+        All transformations are with respect to the local coordinate system.
+    """
+
+    operations = {
+        'union': boolean_union_mesh_mesh,
+        'difference': boolean_difference_mesh_mesh,
+        'intersection': boolean_intersection_mesh_mesh
+    }
+
+    @property
+    def DATASCHEMA(self):
+        import schema
+        return schema.Schema({
+            "attributes": dict,
+            "key": int,
+            "frame": Frame,
+            "shape": Shape,
+            "features": list,
+            "transformations": list,
+        })
+
+    @property
+    def JSONSCHEMANAME(self):
+        return 'part'
+
+    def __init__(self, name, frame=None, shape=None, features=None, **kwargs):
+        super(Part, self).__init__()
+        self._frame = None
+        self.attributes = {'name': name or 'Part'}
+        self.key = None
+        self.frame = frame
+        self.shape = shape or Shape([], [])
+        self.features = features or []
+        self.transformations = deque()
+
+    def __str__(self):
+        tpl = "<Part with shape {} and features {}>"
+        return tpl.format(self.shape, self.features)
+
+    @property
+    def data(self):
+        """dict : A data dict representing the part attributes, the assembly graph identifier, the local coordinate system,
+        the base shape, the shape features, and the transformation tack wrt to the local coordinate system.
+        """
+        data = {
+            'attributes': self.attributes,
+            "key": self.key,
+            "frame": self.frame.data,
+            "shape": self.shape.data,
+            "features": [(shape.data, operation) for shape, operation in self.features],
+            "transformations": [T.data for T in self.transformations],
+        }
+        return data
+
+    @data.setter
+    def data(self, data):
+        self.attributes.update(data['attributes'] or {})
+        self.key = data['key']
+        self.frame.data = data['frame']
+        self.shape.data = data['shape']
+        self.features = [(Shape.from_data(shape), operation) for shape, operation in data['features']]
+        self.transformations = deque([Transformation.from_data(T) for T in data['transformations']])
+
+    @property
+    def frame(self):
+        if not self._frame:
+            self._frame = Frame.worldXY()
+        return self._frame
+
+    @frame.setter
+    def frame(self, frame):
+        self._frame = frame
+
+    @property
+    def geometry(self):
```

Comment:
this method would definitely need memoization of some kind otherwise, it will be recalculated every time the property is accessed even when no changes were made to the shape or features.